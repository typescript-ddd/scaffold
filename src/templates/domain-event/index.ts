export * from "./domain-event.generator";
export * from "./domain-event.template";
export * from "./domain-event.types";
