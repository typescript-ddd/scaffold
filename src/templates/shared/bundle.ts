import { GenerateContext, Chunk } from "./generate-context";
import { Template, TemplateOption } from "./template";

export interface Bundle<
  TValues extends { [K in keyof TOptions]: any },
  TOptions extends { [key: string]: TemplateOption }
> extends Omit<Template<TValues, TOptions>, "generate"> {
  generate(value: TValues, context: GenerateContext): Chunk[];
}
