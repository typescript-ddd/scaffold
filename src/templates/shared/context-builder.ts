import { GenerateContext, ContextImportDeclaration } from "./generate-context";
import { arrays } from "../../utils";
import { SourceFile } from "ts-morph";

export class ContextBuilder implements GenerateContext {
  private _fileName: string;
  private _importDeclarations: ContextImportDeclaration[] = [];

  private constructor(
    private rootDir: string = "@/src",
    fileName: string = "example.ts"
  ) {
    this._fileName = fileName;
  }

  resolveDir(
    type: "domain" | "infrastructure" | "application" | "presentation",
    subPath?: string
  ): string {
    switch (type) {
      case "domain":
        return `${this.rootDir}/domain${subPath ? `/${subPath}` : ""}`;
      case "infrastructure":
        return `${this.rootDir}/infrastructure${subPath ? `/${subPath}` : ""}`;
      case "application":
        return `${this.rootDir}/application${subPath ? `/${subPath}` : ""}`;
      case "presentation":
        return `${this.rootDir}/presentation${subPath ? `/${subPath}` : ""}`;
      default:
        throw new Error(`Unknown type ${type} for resolveDir`);
    }
  }

  getImportDeclarations(moduleSpecifier?: string): ContextImportDeclaration[] {
    const modules = this.getModuleSpecifiers();

    const declarations: ContextImportDeclaration[] = [];

    for (const currentModule of modules) {
      const namedImports = this.getNamedImports(currentModule);
      declarations.push({ moduleSpecifier: currentModule, namedImports });
    }

    if (moduleSpecifier) {
      return declarations.filter((x) => x.moduleSpecifier === moduleSpecifier);
    }

    return declarations;
  }

  addImportDeclaration(declaration: ContextImportDeclaration): void {
    this._importDeclarations.push(declaration);
  }

  getModuleSpecifiers(): string[] {
    return arrays.distinct(
      this._importDeclarations.map((x) => x.moduleSpecifier)
    );
  }

  getNamedImports(moduleSpecifier: string): string[] {
    return arrays.distinct(
      this._importDeclarations
        .filter((x) => x.moduleSpecifier === moduleSpecifier)
        .map((x) => x.namedImports)
        .flat()
    );
  }

  writeSourceFileImports(sourceFile: SourceFile) {
    const importDeclarations = this.getImportDeclarations();
    const moduleSpecifiers = this.getModuleSpecifiers();

    const coreImports = importDeclarations.filter(
      (x) => x.moduleSpecifier === "@typescript-ddd/core"
    );

    if (coreImports) {
      sourceFile.addImportDeclaration({
        moduleSpecifier: `@typescript-ddd/core`,
        namedImports: coreImports.map((x) => x.namedImports).flat(),
      });
    }

    // Non core and root imports
    for (const moduleSpecifier of moduleSpecifiers
      .filter(
        (x) => x !== "@typescript-ddd/core" && !x.startsWith(this.rootDir)
      )
      .sort((a, b) => {
        if (a < b) return -1;
        if (a > b) return 1;
        return 0;
      })) {
      sourceFile.addImportDeclaration({
        moduleSpecifier,
        namedImports: this.getNamedImports(moduleSpecifier),
      });
    }

    // Root imports
    for (const moduleSpecifier of moduleSpecifiers
      .filter((x) => x.startsWith(this.rootDir))
      .sort((a, b) => {
        if (a < b) return -1;
        if (a > b) return 1;
        return 0;
      })) {
      sourceFile.addImportDeclaration({
        moduleSpecifier,
        namedImports: this.getNamedImports(moduleSpecifier),
      });
    }
  }

  clearImportDeclarations(): void {
    this._importDeclarations = [];
  }

  tokenize(text: string, tokens: { name: string }[]): string {
    let output = text;
    tokens.forEach((token) => {
      output = output.replace(`{{${token.name}}}`, "${" + token.name + "}");
    });
    return "`" + output + "`";
  }

  get fileName(): string {
    return this._fileName;
  }

  static build(rootDir: string): ContextBuilder {
    return new ContextBuilder(rootDir);
  }
}
