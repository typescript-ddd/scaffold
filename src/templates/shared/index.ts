export * from "./context-builder";
export * from "./generate-context";