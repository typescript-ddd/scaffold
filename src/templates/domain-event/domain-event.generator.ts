import { Project } from "ts-morph";
import { strings } from "../../utils";
import { Chunk, GenerateContext } from "../shared";
import { DomainEventTemplateValues } from "./domain-event.types";

export const generateDomainEvent = (
  values: DomainEventTemplateValues,
  context: GenerateContext,
  chunkName?: string
): Chunk => {
  const {
    entityName,
    eventAction,
    eventId = `${strings.lower(entityName)}/${strings.lower(eventAction)}`,
  } = values;
  const entityType = strings.capitalize(entityName);
  const actionType = strings.capitalize(eventAction);
  const className = [entityType, actionType, "Event"].join("");
  const entityIdType = `${entityType}Id`;
  const {
    fileName = `${strings.kebab(eventAction, entityName, "event")}.ts`,
    projectPath = ["domain", "event"],
  } = context.currentFile || {};
  const project = new Project();
  const sourceFile = project.createSourceFile(
    fileName,
    `
export class ${className} extends DomainEvent<${entityType}> {}    
`
  );

  context.addImportDeclaration({
    moduleSpecifier: `@typescript-ddd/core`,
    namedImports: ["DomainEvent"],
  });

  context.addImportDeclaration({
    moduleSpecifier: context.resolveDir("domain"),
    namedImports: [entityType, entityIdType],
  });

  context.writeSourceFileImports(sourceFile);

  const classDeclaration = sourceFile.getClass(className);
  classDeclaration.addJsDoc({
    description: `Represents ${strings.describe(className)}.`,
  });

  const constructorDeclaration = classDeclaration.addConstructor({
    parameters: [
      { name: "id", type: entityIdType },
      { name: "payload", type: entityType },
    ],
  });
  constructorDeclaration.setBodyText((writer) => {
    writer.writeLine(`super(id, "${eventId}", payload);`);
  });
  constructorDeclaration.addJsDoc({
    description: `Initializes a new instance of ${strings.describe(
      className
    )}.`,
    tags: [
      {
        tagName: "param",
        text: `{${entityIdType}} id - The identifier of the ${strings.lower(
          entityName
        )}.`,
      },
      {
        tagName: "param",
        text: `{${entityType}} payload - The ${strings.lower(
          entityName
        )} payload.`,
      },
    ],
  });

  return {
    name: chunkName || `Entity${actionType}DomainEvent`,
    content: sourceFile.getFullText(),
    fileName,
    projectPath,
    values,
  };
};
