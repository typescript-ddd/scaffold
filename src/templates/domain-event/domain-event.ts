import { Project, Scope } from "ts-morph";
import { strings } from "../../utils";
import { GenerateContext } from "../shared";

export type GenerateDomainEventOptions = {
  entityName: string;
  eventAction: string;
  eventId?: string;
};

export const generateDomainEvent = (
  options: GenerateDomainEventOptions,
  context: GenerateContext
) => {
  const {
    entityName,
    eventAction,
    eventId = `${strings.lower(entityName)}/${strings.lower(eventAction)}`,
  } = options;
  const entityType = strings.capitalize(entityName);
  const actionType = strings.capitalize(eventAction);
  const className = [entityType, actionType, "Event"].join("");
  const entityIdType = `${entityType}Id`;
  const project = new Project();
  const sourceFile = project.createSourceFile(
    `${context.fileName}.ts`,
    `
export class ${className} extends DomainEvent<${entityType}> {}    
`
  );

  sourceFile.addImportDeclaration({
    moduleSpecifier: `@typescript-ddd/core`,
    namedImports: ["DomainEvent"],
  });
  sourceFile.addImportDeclaration({
    moduleSpecifier: context.resolveDir("domain"),
    namedImports: [entityType, entityIdType],
  });

  const classDeclaration = sourceFile.getClass(className);
  classDeclaration.addJsDoc({
    description: `Represents ${strings.describe(className)}.`,
  });

  const constructorDeclaration = classDeclaration.addConstructor({
    parameters: [
      { name: "id", type: entityIdType },
      { name: "payload", type: entityType },
    ],
  });
  constructorDeclaration.setBodyText((writer) => {
    writer.writeLine(`super(id, "${eventId}", payload);`);
  });
  constructorDeclaration.addJsDoc({
    description: `Initializes a new instance of ${strings.describe(
      className
    )}.`,
    tags: [
      {
        tagName: "param",
        text: `{${entityIdType}} id - The identifier of the ${strings.lower(
          entityName
        )}.`,
      },
      {
        tagName: "param",
        text: `{${entityType}} payload - The ${strings.lower(
          entityName
        )} payload.`,
      },
    ],
  });

  let output = sourceFile.getFullText();

  return output;
};
