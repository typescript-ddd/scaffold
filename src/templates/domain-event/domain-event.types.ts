import { TemplateStringOption } from "../shared/template";

export type DomainEventTemplateValues = {
  entityName: string;
  eventAction: string;
  eventId: string;
};

export type DomainEventTemplateOptions = {
  entityName: TemplateStringOption;
  eventAction: TemplateStringOption;
  eventId: TemplateStringOption;
};
