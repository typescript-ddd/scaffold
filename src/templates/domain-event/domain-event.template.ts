import { GenerateContext } from "../shared";
import { Template } from "../shared/template";
import { generateDomainEvent } from "./domain-event.generator";
import {
  DomainEventTemplateOptions,
  DomainEventTemplateValues,
} from "./domain-event.types";

export class DomainEventTemplate
  implements Template<DomainEventTemplateValues, DomainEventTemplateOptions>
{
  readonly name = "Command";
  readonly description = "Generates a command";
  readonly options: DomainEventTemplateOptions = {
    entityName: {
      name: "entityName",
      label: "Entity Name",
      description: "The name of the entity",
      optionType: "string",
    },
    eventAction: {
      name: "eventAction",
      label: "Event Action",
      description: "The action of the event",
      optionType: "string",
    },
    eventId: {
      name: "eventId",
      label: "Event Id",
      description: "The id of the event",
      optionType: "string",
    },
  };
  readonly defaultValues = {
    entityName: "User",
    eventAction: "Created",
    eventId: "user/created",
  };
  generate(values: DomainEventTemplateValues, context: GenerateContext): string {
    return generateDomainEvent(values, context);
  }
}
