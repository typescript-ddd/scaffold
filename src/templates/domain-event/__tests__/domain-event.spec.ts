import { ContextBuilder, GenerateContext } from "../../shared";
import { generateDomainEvent } from "../domain-event";

describe("DomainEvent", () => {
  let context: GenerateContext;

  beforeEach(() => {
    context = ContextBuilder.build("@src");
  });

  it("Should render template", () => {
    const output = generateDomainEvent(
      { entityName: "User", eventAction: "Created" },
      context
    );
    expect(output).toMatchSnapshot();
  });
});
