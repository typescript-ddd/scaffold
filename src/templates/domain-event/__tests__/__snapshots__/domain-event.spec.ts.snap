// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DomainEvent Should render template 1`] = `
"import { DomainEvent } from "@typescript-ddd/core";
import { User, UserId } from "@src/domain";

/** Represents a user created event. */
export class UserCreatedEvent extends DomainEvent<User> {
    /**
     * Initializes a new instance of a user created event.
     * @param {UserId} id - The identifier of the user.
     * @param {User} payload - The user payload.
     */
    constructor(id: UserId, payload: User) {
        super(id, "user/created", payload);
    }
}    
"
`;
